src/Env.hs

```diff
@@ -0,0 +1,320 @@
+-- | "Env" Contains all the shared code among substance and dsll:
+--   AST, parser and typechecking functions
+--   It also contains the environemt for the typechecker
+--   Author: Dor Ma'ayan, May 2018
+
+{-# OPTIONS_HADDOCK prune #-}
+module Env where
+--module Main (main) where -- for debugging purposes
+
+import Utils
+import System.Process
+import Control.Monad (void)
+import Data.Void
+import Debug.Trace
+import System.IO -- read/write to file
+import System.Environment
+import Control.Arrow ((>>>))
+import System.Random
+import Debug.Trace
+import Data.Functor.Classes
+import Data.List
+import Data.Maybe (fromMaybe)
+import Data.Typeable
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import Text.Megaparsec.Expr
+--import Text.PrettyPrint
+--import Text.PrettyPrint.HughesPJClass hiding (colon, comma, parens, braces)
+import qualified Data.Map.Strict as M
+import qualified Text.Megaparsec.Char.Lexer as L
+
+----------------------------------- AST -----------------------------------------------
+
+data TypeName = TypeNameConst String  -- these are all names, e.g. “Set”
+              | AllT                  -- specifically for global selection in Style
+                deriving (Show, Eq, Typeable)
+
+data TypeVar = TypeVar { typeVarName :: String,
+                         typeVarPos  :: SourcePos }
+               deriving (Show, Typeable)
+
+instance Eq TypeVar where
+  (TypeVar n1 _) == (TypeVar n2 _) = n1 == n2
+
+instance Ord TypeVar where
+  (TypeVar s1 _) `compare` (TypeVar s2 _) = s1 `compare` s2
+
+data Var = VarConst String
+     deriving (Show, Eq, Typeable, Ord)
+
+data Y = TypeVarY TypeVar
+       | VarY Var
+         deriving (Show, Eq, Typeable, Ord)
+
+data T = TTypeVar TypeVar
+       | TConstr ConstructorInvoker
+         deriving (Show, Eq, Typeable)
+
+data ConstructorInvoker = ConstructorInvoker { nameCons :: String,
```

Comment:
I would name this `TypeCtorApp` (type constructor application). Also, don't we need one for applied value constructors? (Their argument types are different)